### ISSUE TITLE
[Bug] ttnn.concat accepts sub_core_grids but silently ignores it on every path except unsharded-input + interleaved-output

### ISSUE THREAD

**Issue by ifabijanic-tt** [tenstorrent/tt-metal#51442]
## Summary
ttnn.concat takes a sub_core_grids argument, but only one of its dispatch paths forwards it. On every other path the argument is accepted, dropped, and the concat program is built on the full Tensix grid. There is no warning and no error at the op level, the caller believes it constrained the op, and the failure only surfaces later (or not at all).

This happens on ttnn/cpp/ttnn/operations/data_movement/concat/concat.cpp

The only path that honours it (line 335):

// For interleaved outputs, if sub_core_grids is provided, use direct path to avoid massaged operations
// which don't currently support sub_core_grids
if (sub_core_grids.has_value() && !first_tensor.is_sharded() &&
    (mem_config.memory_layout() == TensorMemoryLayout::INTERLEAVED)) {
    return ttnn::operations::data_movement::concat_impl(input_tensors, dim, groups, mem_config, sub_core_grids);
}

Anything that misses that condition, sharded inputs, or a non-interleaved output, falls through to the massaged path at the end of the function (line 446):

auto massaged_concat = untilize_rm_retilize_concat.sequence(unsqueeze_squeeze_1D_concat.sequence(non_aligned_last_dim_concat));

const std::vector<ttnn::Tensor>& itensors(input_tensors);
auto res = massaged_concat(itensors, dim, groups);   // <-- sub_core_grids is not passed
massaged_concat has no sub_core_grids parameter at all, and each builder calls concat_impl(..., std::nullopt) internally (lines 89, 137, 182, 255). So the value is dropped, not merely defaulted.

## Reproduction

In models/common/sampling/tt_sampling.py, the compact invalid-vocab tail mask slices the logits, masks only the padded tail, and concats the pieces back. The sampling logits are width-sharded (DECODE_LOGITS_MEMCFG), so first_tensor.is_sharded() is true and the concat takes the massaged path even though sub_core_grids was passed. On a sampling sub-core grid (Llama TG) the reassembly escapes the sampling sub-device.

Current workaround: when a sub-core grid is in use, skip the compact tail mask entirely and fall back to the plain full-width additive mask (one elementwise add, no slice/concat reassembly). See PR #50687.

### Repro sketch

sub_core_grids = ttnn.CoreRangeSet({ttnn.CoreRange(ttnn.CoreCoord(1, 0), ttnn.CoreCoord(3, 3))})

Width-sharded inputs -> first_tensor.is_sharded() is true -> massaged path -> sub_core_grids dropped
a = ttnn.from_torch(..., device=device, layout=ttnn.TILE_LAYOUT, memory_config=<width-sharded cfg>)
b = ttnn.from_torch(..., device=device, layout=ttnn.TILE_LAYOUT, memory_config=<width-sharded cfg>)

out = ttnn.concat([a, b], dim=3, memory_config=a.memory_config(), sub_core_grids=sub_core_grids)
Inspect the resulting program's core ranges: they cover the full Tensix grid, not sub_core_grids.
With a restricted active sub-device this instead fails with
"Kernel group cores do not match sub device cores for programmable core type TENSIX".
